lms/djangoapps/grades/scores.py

```diff
@@ -5,106 +5,242 @@
+
+        Note: r_earned and r_possible are not required when submitting scores
+        via the submissions API.  And so those along with weight (since unused)
+        are invalid and irrelevant.
+
+    csm_scores (ScoresClient):
+
+        The ScoresClient object (already filtered for the user and course),
+        from which a courseware.models.StudentModule object can be retrieved for
+        the block.
+
+        When the score is found from this storage, it implies the user's score
+        for the block was persisted in the Courseware Student Module. Typically,
+        this storage is used for all CAPA problems, including scores calculated
+        by external graders.
+
+        The returned score includes valid values for:
+            r_earned, r_possible - retrieved from CSM
+            w_earned, w_possible - calculated from the raw scores and weight
+            weight, graded - retrieved from the persisted block, if found,
+                else from the latest block content
+
+    persisted_block (.models.BlockRecord):
+        The block values as found in the grades persistence layer. These values
+        are used only if not found from an earlier storage, and take precedence
+        over values stored within the latest content-version of the block.
+
+        When the score is found from this storage, it implies the user has not
+        yet attempted this problem, but the user's grade _was_ persisted.
+
+        The returned score includes valid values for:
+            r_earned - will equal 0.0 since the user's score was not found from
+                earlier storages
+            r_possible - retrieved from the persisted block
+            w_earned, w_possible - calculated from the raw scores and weight
+            weight, graded - retrieved from the persisted block
+
+    block (block_structure.BlockData):
+        Values from the latest content-version of the block are used only if
+        they were not available from a prior storage.
+
+        When the score is found from this storage, it implies the user has not
+        yet attempted this problem and the user's grade was _not_ yet persisted.
+
+        The returned score includes valid values for:
+            r_earned - will equal 0.0 since the user's score was not found from
+                earlier storages
+            r_possible - retrieved from the latest block content
+            w_earned, w_possible - calculated from the raw scores and weight
+            weight, graded - retrieved from the latest block content
+    """
+    weight = _get_weight_from_block(persisted_block, block)
+
+    # Priority order for retrieving the scores:
+    # submissions API -> CSM -> grades persisted block -> latest block content
+    r_earned, r_possible, w_earned, w_possible = (
+        _get_score_from_submissions(submissions_scores, block) or
+        _get_score_from_csm(csm_scores, block, weight) or
+        _get_score_from_block(persisted_block, block, weight)
+    )
```

Comment:
For self-documenting code, instead of adding a comment, I renamed the method to `_get_score_from_persisted_or_latest_block`.  What do you think?